akka-http-core/src/main/scala/akka/http/impl/engine/http2/client/PersistentConnection.scala

```diff
@@ -0,0 +1,127 @@
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = new GraphStageLogic(shape) with StageLogging {
+      become(Unconnected)
+
+      def become(state: State): Unit = setHandlers(requestIn, responseOut, state)
+
+      trait State extends InHandler with OutHandler
+      object Unconnected extends State {
+        override def onPush(): Unit = connect()
+        override def onPull(): Unit =
+          if (!hasBeenPulled(requestIn)) // requestIn might already have been pulled when we failed and went back to Unconnected
+            pull(requestIn)
+
+      }
+
+      def connect(): Unit = {
+        val requestOut = new SubSourceOutlet[HttpRequest]("PersistentConnection.requestOut")
+        val responseIn = new SubSinkInlet[HttpResponse]("PersistentConnection.responseIn")
+
+        Source.fromGraph(requestOut.source)
+          .via(connectionFlow)
+          .runWith(responseIn.sink)(subFusingMaterializer)
+
+        become(new Connected(requestOut, responseIn))
+      }
+
+      class Connected(
+        requestOut: SubSourceOutlet[HttpRequest],
+        responseIn: SubSinkInlet[HttpResponse]
+      ) extends State {
+        private var ongoingRequests: Map[AssociationTag, Map[AttributeKey[_], RequestResponseAssociation]] = Map.empty
+        responseIn.pull()
+
+        requestOut.setHandler(new OutHandler {
+          override def onPull(): Unit =
+            if (!isAvailable(requestIn)) pull(requestIn)
+            else dispatchRequest(grab(requestIn))
+
+          override def onDownstreamFinish(): Unit = onDisconnected()
+        })
+        responseIn.setHandler(new InHandler {
+          override def onPush(): Unit = {
+            val response = responseIn.grab()
+            val tag = response.attribute(associationTagKey).get
+            require(ongoingRequests.contains(tag))
+            ongoingRequests -= tag
+            push(responseOut, response)
+          }
+
+          override def onUpstreamFinish(): Unit = onDisconnected()
+          override def onUpstreamFailure(ex: Throwable): Unit = onDisconnected() // FIXME: log error
+        })
+        def onDisconnected(): Unit = {
+          emitMultiple[HttpResponse](responseOut, ongoingRequests.values.map(errorResponse.withAttributes(_)).toVector, () => setHandler(responseOut, Unconnected))
+          responseIn.cancel()
+          requestOut.fail(new RuntimeException("connection broken"))
+          // become(Unconnected) doesn't work because of using emit
+          // so we need to do it more carefully here
+          setHandler(requestIn, Unconnected)
+          if (isAvailable(responseOut) && !hasBeenPulled(requestIn)) pull(requestIn)
+        }
+
+        def dispatchRequest(req: HttpRequest): Unit = {
+          val tag = new AssociationTag
+          ongoingRequests += (tag -> req.attributes.collect {
+            case (key, value: RequestResponseAssociation) => key -> value
+          })
```

Comment:
Could there be more than one?

Reply:
It's already collecting all `RequestResponseAssociation` values. Is that what you mean?

Reply:
I don't even remember what I meant anymore! :D